lib/readline.js

```diff
@@ -965,91 +975,124 @@ const escapeCodeReAnywhere = new RegExp([
   functionKeyCodeReAnywhere.source, metaKeyCodeReAnywhere.source, /\x1b./.source
 ].join('|'));
 
-function emitKeys(stream, s) {
-  if (s instanceof Buffer) {
-    if (s[0] > 127 && s[1] === undefined) {
-      s[0] -= 128;
-      s = '\x1b' + s.toString(stream.encoding || 'utf-8');
-    } else {
-      s = s.toString(stream.encoding || 'utf-8');
-    }
-  }
-
-  var buffer = [];
-  var match;
-  while (match = escapeCodeReAnywhere.exec(s)) {
-    buffer = buffer.concat(s.slice(0, match.index).split(''));
-    buffer.push(match[0]);
-    s = s.slice(match.index + match[0].length);
-  }
-  buffer = buffer.concat(s.split(''));
-
-  buffer.forEach(function(s) {
-    var ch,
-        key = {
-          sequence: s,
-          name: undefined,
-          ctrl: false,
-          meta: false,
-          shift: false
-        },
-        parts;
-
-    if (s === '\r') {
-      // carriage return
-      key.name = 'return';
 
-    } else if (s === '\n') {
-      // enter, should have been called linefeed
-      key.name = 'enter';
+function* emitKeys(stream) {
+  while (true) {
+    var ch = yield;
+    var s = ch;
+    var escaped = false;
+    var key = {
+      sequence: null,
+      name: undefined,
+      ctrl: false,
+      meta: false,
+      shift: false
+    };
+
+    if (ch === '\x1b') {
+      escaped = true;
+      s += (ch = yield);
+
+      if (ch === '\x1b') {
+        s += (ch = yield);
+      }
+    }
 
-    } else if (s === '\t') {
-      // tab
-      key.name = 'tab';
+    if (escaped && (ch === 'O' || ch === '[')) {
+      // ansi escape sequence
+      var code = ch;
+      var modifier = 0;
 
-    } else if (s === '\b' || s === '\x7f' ||
-               s === '\x1b\x7f' || s === '\x1b\b') {
-      // backspace or ctrl+h
-      key.name = 'backspace';
-      key.meta = (s.charAt(0) === '\x1b');
+      if (ch === 'O') {
+        // ESC O letter
+        // ESC O modifier letter
+        s += (ch = yield);
 
-    } else if (s === '\x1b' || s === '\x1b\x1b') {
-      // escape key
-      key.name = 'escape';
-      key.meta = (s.length === 2);
+        if (ch >= '0' && ch <= '9') {
+          modifier = parseInt(ch, 10) - 1;
+          s += (ch = yield);
+        }
 
-    } else if (s === ' ' || s === '\x1b ') {
-      key.name = 'space';
-      key.meta = (s.length === 2);
+        code += ch;
 
-    } else if (s.length === 1 && s <= '\x1a') {
-      // ctrl+letter
-      key.name = String.fromCharCode(s.charCodeAt(0) + 'a'.charCodeAt(0) - 1);
-      key.ctrl = true;
+      } else if (ch === '[') {
```

Comment:
I don't know any escape sequences that have three brackets (like `\x1b[[[1~`), and io.js can't parse them (there is a whitelist of accepted codes in the code below).

So if we add `while` here, you can press `ESC`, and every single `[` character after that will be silently ignored, because parser will be still waiting thinking it's in the middle of an escape sequence.

You know that "WAT" effect when you press ESC in the bash for example, and several characters after that are getting ignored by the terminal. While loop will make the same effect only worse.

So I prefer to be deterministic here. Right now, this parser is guaranteed to emit at least one keypress event every 9 characters (something like `\x1b\x1b[[12;3~` is the longest sequence). That's why there is no while loop there.

Same thing with digits below.